### PR TITLE
Focuser

### PR DESCRIPTION
**Pull request by sglushkov**
Focuser branch from sglushkov.
Added temperature compensation features. Fo user temperature, focuser auto compensation controls.

### CODE REVIEW

File: ain_imager_src/handlepropertychange.cpp [hunk 1]
```diff
@@ -717,6 +717,54 @@ void update_cooler_power(ImagerWindow *w, indigo_property *property) {
...
 
+void update_focuser_temperature(ImagerWindow *w, indigo_property *property) {
+	indigo_debug("change %s", property->name);
+	for (int i = 0; i < property->count; i++) {
+		if (client_match_item(&property->items[i], FOCUSER_TEMPERATURE_ITEM_NAME)) {
+			indigo_debug("change %s = %f", property->items[i].name, property->items[i].number.value);
+			char temperature[INDIGO_VALUE_SIZE];
+			snprintf(temperature, INDIGO_VALUE_SIZE, "%.2f%", property->items[i].number.value);
+			w->set_text(w->m_focuser_temp, temperature);
+			w->set_text(w->m_focuser_temp, temperature);
```
Comment (rumengb) [inline ain_imager_src/handlepropertychange.cpp]: no need to be two times

File: ain_imager_src/changeproperty.cpp [hunk 2]
```diff
@@ -137,6 +137,15 @@ void ImagerWindow::change_agent_gain_property(const char *agent, QSpinBox *ccd_g
 	indigo_change_number_property_1(nullptr, agent, CCD_GAIN_PROPERTY_NAME, CCD_GAIN_ITEM_NAME, value);
 }
 
+void ImagerWindow::change_agent_temperature_compensation_steps(const char *agent) const {
+	static const char *items[] = {
+		FOCUSER_COMPENSATION_ITEM_NAME
+		};
```
Comment (rumengb) [inline ain_imager_src/changeproperty.cpp]: please align brackets

File: ain_imager_src/changeproperty.cpp [hunk 3]
```diff
@@ -137,6 +137,15 @@ void ImagerWindow::change_agent_gain_property(const char *agent, QSpinBox *ccd_g
...
+	static double values[1];
+	values[0] = (double)m_focuser_temperature_compensation_steps->value();
+	indigo_change_number_property(nullptr, agent, FOCUSER_COMPENSATION_PROPERTY_NAME, 1, items, values);
```
Comment (rumengb) [inline ain_imager_src/changeproperty.cpp]: better use indigo_change_number_property_1() finction

File: ain_imager_src/focuser_tab.cpp [hunk 4]
```diff
@@ -383,6 +404,45 @@ void ImagerWindow::create_focuser_tab(QFrame *focuser_frame) {
...
+	label = new QLabel("Steps per degree:");
+	temperature_compensation_steps_frame_layout->addWidget(label, temp_comp_steps_row, 0, 1, 3);
+	m_focuser_temperature_compensation_steps = new QSpinBox();
+	m_focuser_temperature_compensation_steps->setMaximum(10000);
+	m_focuser_temperature_compensation_steps->setMinimum(-10000);
+	m_focuser_temperature_compensation_steps->setValue(0);
+	m_focuser_temperature_compensation_steps->setEnabled(false);
+	m_focuser_temperature_compensation_steps->setKeyboardTracking(false);
+	temperature_compensation_steps_frame_layout->addWidget(m_focuser_temperature_compensation_steps, temp_comp_steps_row, 3);
+
```
Comment (rumengb) [inline ain_imager_src/focuser_tab.cpp:438]: remove this emty line

File: ain_imager_src/focuser_tab.cpp [hunk 5]
```diff
@@ -517,7 +600,6 @@ void ImagerWindow::on_image_right_click(double x, double y) {
 	});
 }
 
-
```
Comment (rumengb) [inline ain_imager_src/focuser_tab.cpp:520]: what changed here?

**Review by rumengb — CHANGES_REQUESTED** [review]

File: ain_imager_src/focuser_tab.cpp (hunk 4 above)
Comment (sglushkov) [reply ain_imager_src/focuser_tab.cpp:438]: Removed

File: ain_imager_src/changeproperty.cpp (hunk 3 above)
Comment (sglushkov) [reply ain_imager_src/changeproperty.cpp]: OK. Changed to _1

File: ain_imager_src/changeproperty.cpp (hunk 2 above)
Comment (sglushkov) [reply ain_imager_src/changeproperty.cpp]: Aligned.

File: ain_imager_src/handlepropertychange.cpp (hunk 1 above)
Comment (sglushkov) [reply ain_imager_src/handlepropertychange.cpp]: OK.Deleted.

**Pull request merged**